### PR TITLE
Intake

### PR DESCRIPTION
**Pull request by nmot919**
Adds the intake subsystem, also adds intake class to robot map with functions for it. And the gaston map which contains the pid coefficients for the deployment motor.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -4,18 +4,24 @@
   // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
   "version": "0.2.0",
   "configurations": [
-
+    {
```
Comment (msoucy) [inline .vscode/launch.json:7]: Why was this added? The two below should cover simulation.

File: src/main/java/frc/robot/Robot.java [hunk 2]
```diff
@@ -1,29 +1,36 @@
...
   public void robotInit() {
-    // Nothing to initialize
+    super.robotInit();
   }
 
   @Override
   public void configureButtonBindings() {
-    // TODO Auto-generated method stub
-
+    controller.getAButton();
```
Comment (msoucy) [inline src/main/java/frc/robot/Robot.java]: What is this doing?

File: src/main/java/frc/robot/maps/RobotMap.java [hunk 3]
```diff
@@ -1,9 +1,60 @@
...
+            return insideLimit;
+        }
+
+        public BooleanSupplier getOutsideLimit() {
+            return outsideLimit;
+        }
+    }
 
+    public IntakeMap getIntakeMap() {
+        return new IntakeMap(new SmartMotorController(), new SmartMotorController(), new MockDigitalInput(),
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/RobotMap.java]: Why not use the default constructor?

File: src/main/java/frc/robot/maps/GastonMap.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        deploymentPidController.setIZone(IZone);
+        deploymentPidController.setFF(feedForward);
+        deploymentPidController.setOutputRange(minOutput, maxOutput);
+
+        deploymentPidController.setSmartMotionMaxVelocity(smartMotionMaxVel, 0);
+        deploymentPidController.setSmartMotionMinOutputVelocity(smartMotionOutputVel, 0);
+        deploymentPidController.setSmartMotionMaxAccel(smartMotionMaxAccel, 0);
+
+        // the value here is in rotations
+        deploymentPidController.setReference(0.25, CANSparkMax.ControlType.kSmartMotion);
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/GastonMap.java]: Do we want to set the reference here? Doesn't the reference change during operation?

If so, then when you turn the robot on it'll tell the robot to put the reference at that position... so when it gets enabled you get a surprise movement.

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                () -> {
+                    rollerMotor.set(s.get(rollerSpeed));
+                }, () -> {
+                    rollerMotor.set(0);
+                });
+    }
+
+    public CommandBase retractIntake() {
+        return cmd("Retract Intake").onExecute(() -> {
+            if (!insideLimit.getAsBoolean()) {
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Intake.java]: Here and and in deploy, look at the Modifier concept in chopshoplib (used in the climber branch) to make it more clear that you're limiting movement - right now there's a pretty major bug

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+    @Override
+    public void periodic() {
+        // This method will be called once per scheduler run
+        SmartDashboard.putBoolean("Intake Deployed", !insideLimit.getAsBoolean());
+    }
+
+    @Override
+    public void safeState() {
+
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Intake.java:66]: What is a safe state for this subsystem?

**Review by msoucy — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Robot.java [hunk 7]
```diff
@@ -1,29 +1,38 @@
...
-    // Nothing to initialize
+    super.robotInit();
   }
 
   @Override
   public void configureButtonBindings() {
-    // TODO Auto-generated method stub
-
+    controller.a().whileHeld(parallel("Deploy and Start Roller", intake.setRollerSpeed(SpinDirection.COUNTERCLOCKWISE),
+        intake.deployIntake())).whenReleased(intake.retractIntake());
```
Comment (bot190) [inline src/main/java/frc/robot/Robot.java]: Generally we create a command group inside the `intake` class that handles doing these two things so we don't have to deal with it at the robot level.

File: src/main/java/frc/robot/maps/GastonMap.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    private final DigitalInput outsideLimit = new DigitalInput(4);
+    private final DigitalInput insideLimit = new DigitalInput(5);
+
+    // initializes relative encoder and pid controller, we don't need the encoder rn
+
+    private PIDSparkMax deploymentMotor = new PIDSparkMax(2, MotorType.kBrushless);
+
+    // private RelativeEncoder deploymentEncoder =
+    // deploymentMotor.getEncoder().getRaw();
+    private final SparkMaxPIDController deploymentPidController = deploymentMotor.getPidController();
```
Comment (bot190) [inline src/main/java/frc/robot/maps/GastonMap.java]: All of these should be defined inside of `getIntakeMap`, no need to have them at the class level.

File: src/main/java/frc/robot/maps/GastonMap.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+    public IntakeMap getIntakeMap() {
+        // PID coefficients
+
+        double P = 5e-5;
+        double I = 1e-6;
```
Comment (bot190) [inline src/main/java/frc/robot/maps/GastonMap.java]: These can be 0 for now, unless you've calculated these based on anything specific.

File: src/main/java/frc/robot/maps/GastonMap.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        double smartMotionOutputVel = 0;
+        double smartMotionMaxAccel = 600;
+
+        // TODO pid/smart motion coefficients for intake
+
+        deploymentPidController.setP(P);
+        deploymentPidController.setI(I);
+        deploymentPidController.setD(D);
+        deploymentPidController.setIZone(IZone);
+        deploymentPidController.setFF(feedForward);
```
Comment (bot190) [inline src/main/java/frc/robot/maps/GastonMap.java]: We generally don't need FF for position PID, no need to set it.

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+public class Intake extends SmartSubsystemBase {
+
+    private final SmartMotorController deploymentMotor;
+    private final SmartMotorController rollerMotor;
+
+    private final BooleanSupplier insideLimit;
+    private final BooleanSupplier outsideLimit;
+
+    private final double rollerSpeed = 1;
+    private final double deploySpeed = 1;
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Intake.java]: These should be `private final static double` and the naming should be in upper snake case.

See https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names for more info.

```suggestion
    private static final double ROLLER_SPEED= 1;
    private static final double DEPLOY_SPEED= 1;
```

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    private final Modifier insideModifier;
+    private final Modifier outsideModifier;
+
+    public Intake(final IntakeMap map) {
+        this.deploymentMotor = map.getDeploy();
+        this.rollerMotor = map.getRoller();
+        this.insideLimit = map.getInsideLimit();
+        this.outsideLimit = map.getOutsideLimit();
+        outsideModifier = Modifier.upperLimit(outsideLimit);
+        insideModifier = Modifier.lowerLimit(insideLimit);
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Intake.java]: Probably worth prefixing with `this` for consistency sake.

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        // :) we don't appreciate
+        // deploymentMotor.setControlType(CANSparkMax.ControlType.kSmartMotion);
+    }
+
+    // Counterclockwise ball go in, clockwise ball go out
+    public CommandBase setRollerSpeed(SpinDirection rDirection) {
```
Comment (bot190) [inline src/main/java/frc/robot/subsystems/Intake.java]: what is `r`? It's worth spelling out `rollerDirection`.

**Review by bot190 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+    public Intake(final IntakeMap map) {
+        this.deploymentMotor = map.getDeploy();
+        this.rollerMotor = map.getRoller();
+        this.insideLimit = map.getInsideLimit();
+        this.outsideLimit = map.getOutsideLimit();
+        outsideModifier = Modifier.upperLimit(outsideLimit);
+        insideModifier = Modifier.lowerLimit(insideLimit);
+        // :) we don't appreciate
+        // deploymentMotor.setControlType(CANSparkMax.ControlType.kSmartMotion);
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Intake.java]: You can now use `PIDControlType.SmartMotion` for this

File: src/main/java/frc/robot/maps/RobotMap.java [hunk 15]
```diff
@@ -1,9 +1,60 @@
...
 public class RobotMap {
 
+    public static class IntakeMap {
+        private final SmartMotorController deploymentMotor;
+        private final SmartMotorController rollerMotor;
+        private final BooleanSupplier insideLimit;
+        private final BooleanSupplier outsideLimit;
+
+        public IntakeMap() {
+            this.deploymentMotor = new SmartMotorController();
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/RobotMap.java]: You can call `this(new SmartMotorController(), ...)` instead of setting each item separately, if it's easier to read

File: src/main/java/frc/robot/maps/GastonMap.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+        deploymentPidController.setSmartMotionMaxVelocity(smartMotionMaxVel, 0);
+        deploymentPidController.setSmartMotionMinOutputVelocity(smartMotionOutputVel, 0);
+        deploymentPidController.setSmartMotionMaxAccel(smartMotionMaxAccel, 0);
+
+        // we dont do this here
+        // the value here is in rotations
+        // deploymentPidController.setReference(0.25,
+        // CANSparkMax.ControlType.kSmartMotion);
+
+        IntakeMap map = new IntakeMap(new PIDSparkMax(3, MotorType.kBrushless),
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/GastonMap.java]: Temporary variable that serves no real purpose

File: src/main/java/frc/robot/Robot.java (hunk 7 above)
Comment (msoucy) [reply src/main/java/frc/robot/Robot.java]: As discussed on Wednesday, this also will get funky when you're trying to run two things in parallel from the same subsystem.

**Review by msoucy — CHANGES_REQUESTED** [review]

**Comment by msoucy** [comment]
In addition to the existing comments, you'll want to merge `main` into this branch and resolve the merge conflicts.

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    public Intake(final IntakeMap map) {
+        this.deploymentMotor = map.getDeploy();
+        this.rollerMotor = map.getRoller();
+        this.insideLimit = map.getInsideLimit();
+        this.outsideLimit = map.getOutsideLimit();
+        this.outsideModifier = Modifier.upperLimit(outsideLimit);
+        this.insideModifier = Modifier.lowerLimit(insideLimit);
+
+        deploymentMotor.setControlType(PIDControlType.SmartMotion);
+        deploymentMotor.setSetpoint(0.25);
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Intake.java]: Probably shouldn't set this in the constructor - beware of wild movement

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    // Counterclockwise ball go in, clockwise ball go out
+
+    public CommandBase startIntakeMechanism(SpinDirection rollerDirection) {
```
Comment (msoucy) [inline src/main/java/frc/robot/subsystems/Intake.java]: Bit of a misnomer, how about `runMechanism`?

File: src/main/java/frc/robot/maps/GastonMap.java [hunk 19]
```diff
@@ -17,11 +18,57 @@
...
+        deploymentPidController.setI(I);
+        deploymentPidController.setD(D);
+        deploymentPidController.setIZone(IZone);
+        deploymentPidController.setOutputRange(minOutput, maxOutput);
+
+        deploymentPidController.setSmartMotionMaxVelocity(smartMotionMaxVel, 0);
+        deploymentPidController.setSmartMotionMinOutputVelocity(smartMotionOutputVel, 0);
+        deploymentPidController.setSmartMotionMaxAccel(smartMotionMaxAccel, 0);
+
+        IntakeMap map = new IntakeMap(rollerMotor, deploymentMotor, outsideLimit::get, insideLimit::get);
```
Comment (msoucy) [inline src/main/java/frc/robot/maps/GastonMap.java]: Unnecessary temporary variable

**Review by msoucy — CHANGES_REQUESTED** [review]
Looks good, just a few minor tweaks left.

File: src/main/java/frc/robot/subsystems/Intake.java (hunk 17 above)
Comment (nmot919) [reply src/main/java/frc/robot/subsystems/Intake.java]: where would a good place for it be?

Comment (msoucy) [reply src/main/java/frc/robot/subsystems/Intake.java]: Instead of `deploymentMotor.set(xyz)` you'd use `setSetpoint(xyz)` - you're setting a position to move to, not a speed

**Review by msoucy — DISMISSED** [review]
Merging so we have the infrastructure, but I just realized some issues with the commands that need addressing.

Basically, what I noticed is that the modifiers don't really work when using `setSetpoint`. So what probably makes the most sense, is to have two commands.

The default command would be a `running` command, that checks if the limit is pressed. If not, set the setpoint to the right value. If so, stop it.

The "activate" command would be a `running` command that does the opposite.

**Review by msoucy — APPROVED** [review]

File: src/main/java/frc/robot/maps/GastonMap.java [hunk 20]
```diff
@@ -17,11 +18,49 @@
     private final PIDSparkMax leftMotor = new PIDSparkMax(0, MotorType.kBrushless);
     private final PIDSparkMax rightMotor = new PIDSparkMax(1, MotorType.kBrushless);
 
+    // intake stuff
+    private final DigitalInput outsideLimit = new DigitalInput(4);
+    private final DigitalInput insideLimit = new DigitalInput(5);
+
+    private final PIDSparkMax deploymentMotor = new PIDSparkMax(2, MotorType.kBrushless);
+    private final PIDSparkMax rollerMotor = new PIDSparkMax(3, MotorType.kBrushless);
+    private final SparkMaxPIDController deploymentPidController = deploymentMotor.getPidController();
```
Comment (bot190) [inline src/main/java/frc/robot/maps/GastonMap.java:27]: Please put these inside `getIntakeMap`

**Review by bot190 — APPROVED** [review]
Since the intake deployment is far more straightforward now we likely don't need to use PID at all for it. One thing we may want to be able to do is retract the intake while still running the roller counterclockwise. This should help intake any ball that the intake started to pull in but hadn't made it all the way.
